Add ComWrappers tutorial (#25933)
dotnet/docs: docs/standard/native-interop/tutorial-comwrappers.md

---
title: "Using the ComWrappers API"
description: A tutorial on properly implementing the ComWrappers API.
author: AaronRobinsonMSFT
ms.date: "09/02/2021"
helpviewer_keywords:
  - "COM interop, ComWrappers"
  - "RCW"
  - "CCW"
  - "interoperation with unmanaged code, COM wrappers"
  - "NativeAOT"
---
# Tutorial: Use the `ComWrappers` API

In this tutorial, you'll learn how to properly subclass the [`ComWrappers`][api_comwrappers] type to provide an optimized and AOT-friendly COM interop solution. Before starting this tutorial, you should be familiar with COM, its architecture, and existing COM interop solutions.

In this tutorial, you'll implement the following interface definitions. These interfaces and their implementations will demonstrate:

- Marshalling and unmarshalling types across the COM/.NET boundary.
- Two distinct approaches to consuming native COM objects in .NET.
- A recommended pattern for enabling custom COM interop in .NET 5 and beyond.

All source code used in this tutorial is available in the [dotnet/samples repository](https://github.com/dotnet/samples/tree/main/core/interop/comwrappers/Tutorial).

**C# definitions**

```csharp

> interface IDemoGetType

Author: Oh boy. This is meant to be `IDemoGet` type rather than "get type".

We won't implement `ReleaseObjects()` in this tutorial.

### Step 2 &ndash; Implement `ComputeVtables()`

Let's start with the Managed Object Wrapper &ndash; these wrappers are easier. You'll build a [Virtual Method Table][wiki_vtable], or *vtable*, for each interface in order to project them into the COM environment. For this tutorial, you'll define a vtable as a sequence of pointers, where each pointer represents an implementation of a function on an interface &ndash; order is _very_ important here. In COM, every interface inherits from [`IUnknown`][api_iunknown]. The `IUnknown` type has three methods defined in the following order: `QueryInterface()`, `AddRef()`, and `Release()`. After the `IUnknown` methods come the specific interface methods. For example, consider `IDemoGetType` and `IDemoStoreType`. Conceptually, the vtables for the types would look like the following:

```
IDemoGetType    | IDemoStoreType
==================================
QueryInterface  | QueryInterface
AddRef          | AddRef
Release         | Release
GetString       | StoreString
```

> Looking at `DemoImpl`, we already have an implementation for `GetString()` and `StoreString()`, but what about the `IUnknown` functions? How to [implement an `IUnknown` instance][doc_impliunknown] is beyond the scope of this tutorial, but it can be done manually in `ComWrappers`. However, in this tutorial, you'll let the runtime handle that part. You can get the `IUnknown` implementation using the [`ComWrappers.GetIUnknownImpl()`][api_comwrappers] method.

Reviewer: Is it intentional that the link for `GetIUnknownImpl` points to the more generic `ComWrappers` reference, why not the method page itself?

Reviewer: Another piece of feedback is on the API reference for this method - it says very little. Would be nice to have some description of what the runtime implementation does (and roughly how) - specifically the QI behavior is interesting I think.

Author: > description of what the runtime implementation does

That is fair. I will add some notes.

> (and roughly how)

The how is an implementation detail I'd rather not document as it is always subject to change. Also, it shouldn't be relevant in the majority of cases.

Reviewer: > The how is an implementation detail I'd rather not document as it is always subject to change. Also, it shouldn't be relevant in the majority of cases.

I completely agree that it's an implementation detail - I was just curious :-)

Author: It is just a linear search in two sets of arrays. The implementation is relatively simple and documented: https://github.com/dotnet/runtime/blob/404a89f843b121dfb3909c645eda80cc3a2aa9cc/src/coreclr/interop/comwrappers.cpp#L370-L456. So much easier than the built-in system...

```csharp
[UnmanagedCallersOnly]
public static int GetString(IntPtr _this, IntPtr* str);
```

The wrapper implementation of `IDemoGetType.GetString()` should consist of marshalling logic and then a dispatch to the managed object being wrapped. All the state for dispatch is contained within the provided `_this` argument. The `_this` argument will actually be of type `ComInterfaceDispatch*`. This type represents a low-level structure with a single field, `Vtable`, that will be discussed later. Further details of this type and its layout are an implementation detail of the runtime and should not be depended upon. In order to retrieve the managed instance from a `ComInterfaceDispatch*` instance, use the following code:

```csharp

> IDemoGetType inst = ComInterfaceDispatch.GetInstance<IDemoGetType>((ComInterfaceDispatch*)_this);

Author: I prefer the cast and being explicit here since just accepting an `IntPtr`, in my opinion, is a bit too much freedom.

```

Now that you have a C# method that can be inserted into a vtable, you can construct the vtable. Note the use of [`RuntimeHelpers.AllocateTypeAssociatedMemory()`][api_allocatetypeassociatedmemory] for allocating memory in a way that works with [unloadable][doc_unloadability] assemblies.

```csharp
GetIUnknownImpl(
    out IntPtr fpQueryInterface,
    out IntPtr fpAddRef,
    out IntPtr fpRelease);

> {

Reviewer: I would recommend somehow visually separating the above call. At first glance it looks like a method definition and body for `GetIUnknownImpl`  (especially since the inline variable declaration looks exactly like parameter definition). The only syntax difference is the semicolon. (I would say this is C#'s fault, but that doesn't solve anything).

    int tableCount = 4;
    int idx = 0;
    var vtable = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(
        typeof(DemoComWrappers),
        IntPtr.Size * tableCount);
    vtable[idx++] = fpQueryInterface;
    vtable[idx++] = fpAddRef;
    vtable[idx++] = fpRelease;

>     vtable[idx++] = (IntPtr)(delegate* unmanaged<IntPtr, IntPtr*, int>)&ABI.

Reviewer: This is more of a C# rant - why is it necessary to type out the "native delegate" type here? The funny thing is that the compiler knows exactly what it should be, since if I write it wrong it will complain. It would be nice if just `(IntPtr)&Method` worked.

Author: We should take this offline. There is actually a reason and it relates to our calling convention extension model.

IDemoGetTypeManagedWrapper.GetString;
    Debug.Assert(tableCount == idx);
    s_IDemoGetTypeVTable = (IntPtr)vtable;
}
```

> The allocation of vtables is the first part of implementing `ComputeVtables()`. You should also construct comprehensive definitions for types that you're planning to support &ndash; think `DemoImpl`. Using the constructed vtables, you can now create a series of [`ComInterfaceEntry`][api_cominterfaceentry] instances that represent the complete view of the managed object in COM.

Author: "COM View" is a good idea.

```csharp
{
    s_DemoImplDefinitionLen = 2;
    int idx = 0;
    var entries = (ComInterfaceEntry*)RuntimeHelpers.AllocateTypeAssociatedMemory(
        typeof(DemoComWrappers),
        sizeof(ComInterfaceEntry) * s_DemoImplDefinitionLen);
    entries[idx].IID = IDemoGetType.IID_IDemoGetType;
    entries[idx++].Vtable = s_IDemoGetTypeVTable;
    entries[idx].IID = IDemoStoreType.IID_IDemoStoreType;
    entries[idx++].Vtable = s_IDemoStoreVTable;
    Debug.Assert(s_DemoImplDefinitionLen == idx);
    s_DemoImplDefinition = entries;
}

> ```

Reviewer: Comment on the full code in the samples repo:
```C#
Debug.Assert(flags.HasFlag(CreateComInterfaceFlags.None));
```

The condition is always true since `None == 0`. It should probably be `Debug.Assert(flags is CreateComInterfaceFlags.None)`.

The allocation of vtables and entries for the Managed Object Wrapper can and should be done ahead of time since the data can be used for all instances of the type. The work here could be performed in a `static` constructor or a module initializer, but it should be done ahead of time so the `ComputeVtables()` method is as simple and quick as possible.

> 

Author: That is fair. In this case those I would argue the types the ComWrappers implementation supports should be left alone because you never know what type will appear from COM.

> 

Reviewer: > the types the ComWrappers implementation supports should be left alone because you never know what type will appear from COM.

That is absolutely true - but what if the code which uses the ComWrappers is not used... in that case the entire ComWrappers class can be removed. Unless it's rooted somewhere else, like the module constructor.
It's also at odds with "pay for play" - the app would pay the price of initializing it up front, without knowing if it will need it or not.

> 

Author: >It's also at odds with "pay for play" - the app would pay the price of initializing it up front, without knowing if it will need it or not.

I would like to think more about that UX. The `ComWrappers` system is going to include substantial code gen, so why generate it in the first place if it isn't going to be needed? In fact that seems like a layering issue rather than push something so big into the module initializer vs a `static` type initializer. Perhaps a compromise here would be to add a comment that states module initializers may not be the best approach for trimmer support?

```csharp

> protected override unsafe ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, 

Author: > If I implement my own QI - can I just register IUnknown and nothing else here? 

Yep, that is an option. More advanced than even what we have here.

>I assume the runtime provided QI implementation uses the returned table to do its job, right?

Right.

> Is it valid (expected) to return different interface set per-instance (even of the same type)?

That is possible for sure but it would make things difficult to follow and possibly fail in some COM scenarios depending on how the type is exposed.

out int count)
{
    if (obj is DemoImpl)
    {
        count = s_DemoImplDefinitionLen;
        return s_DemoImplDefinition;
    }

    // Unknown type
    count = 0;
    return null;
}
```

> Once you've implemented the `ComputeVtables()` method, the `ComWrappers` subclass will be able to produce Managed Object Wrappers for instances of `DemoImpl`. Be aware that the returned Managed Object Wrapper from the call to [`GetOrCreateComInterfaceForObject()`](xref:System.Runtime.InteropServices.ComWrappers.GetOrCreateComInterfaceForObject%2A) is of type `IUnknown*`. If the native API that's being passed to the wrapper requires a different interface, a [`Marshal.QueryInterface()`][api_marshalqueryinterface] for that interface must be performed.

Author: > Will the runtime construct such a vtable?

Yes, by default. Users can provide their own though.

```csharp
var cw = new DemoComWrappers();
var demo = new DemoImpl();

> IntPtr ccw = cw.GetOrCreateComInterfaceForObject(demo, CreateComInterfaceFlags.None);

Author: I like using the `ccw` and `rcw` for the variables as it is familiar for some.

> IntPtr ccw = cw.GetOrCreateComInterfaceForObject(demo, CreateComInterfaceFlags.None);

Author: Sure.

#### Static Native Object Wrapper

Let's look at the static implementation first. The static Native Object Wrapper involves defining a managed type that implements the .NET interfaces and can forward the calls on the managed type to the COM instance. A rough outline of the static wrapper follows.

```csharp
class DemoNativeStaticWrapper
    : IDemoGetType
    , IDemoStoreType
{
    public string? GetString() =>
        throw new NotImplementedException();

    public void StoreString(int len, string? str) =>
        throw new NotImplementedException();
}
```

To construct an instance of this class and provide it as a wrapper, you must define some policy. If this type is used as a wrapper, it would seem that since it implements both interfaces, the underlying COM instance should implement both interfaces too. Given that you're adopting this policy, you'll need to confirm this through calls to [`Marshal.QueryInterface()`][api_marshalqueryinterface] on the COM instance.

```csharp

> int hr = Marshal.QueryInterface(ptr, ref IDemoGetType.IID_IDemoGetType, out IntPtr IDemoGetTypeInst);

Reviewer: I guess a C# rant - the `IID_IDemoGetType` can't be a readonly field - since the "ref" here is "in/out".

Author: I know, I know...

if (hr != 0)
{
    return null;
}

hr = Marshal.QueryInterface(ptr, ref IDemoStoreType.IID_IDemoStoreType, out IntPtr IDemoStoreTypeInst);
if (hr != 0)
{

>     Marshal.Release(ptr);

Author: You are absolutely right. This is a failure on my part.

    return null;
}

return new DemoNativeStaticWrapper()
{
    IDemoGetTypeInst = IDemoGetTypeInst,
    IDemoStoreTypeInst = IDemoStoreTypeInst
};
```

#### Dynamic Native Object Wrapper

> Dynamic wrappers are more flexible because they provide a way for types to be queried at run time instead of statically. In order to provide this support, you'll utilize [`IDynamicInterfaceCastable`][api_idynamicinterfacecastable]. Observe that `DemoNativeDynamicWrapper` only implements this interface. The functionality that the interface provides is a chance to determine what type is supported at run time.

Reviewer: > types to be queried at run time instead of statically.

I think this is confusing. Even the static one queries the types at runtime. If I understand this correctly the core difference is that with the static one all of the COM interfaces and their managed counterparts must be figured out up front (for each instance it's needed on), the dynamic one is lazy, it only needs to figure out the managed interface when something actually asks for it.

Curiosity: How does the `IDynamicallyInterfaceCastable` play with reflection - for example what happens if I call `obj.GetType().GetInterfaces()` - will it include the dynamic interfaces or not?

        throw new NotImplementedException();

    public bool IsInterfaceImplemented(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented) =>
        throw new NotImplementedException();
}
```

Let's look at one of the interfaces that `DemoNativeDynamicWrapper` will dynamically support. The following code provides the implementation of `IDemoStoreType` using the *default interface methods* feature.

```csharp
[DynamicInterfaceCastableImplementation]
unsafe interface IDemoStoreTypeNativeWrapper : IDemoStoreType
{
    public static void StoreString(IntPtr inst, int len, string? str);

    void IDemoStoreType.StoreString(int len, string? str)
    {
        var inst = ((DemoNativeDynamicWrapper)this).IDemoStoreTypeInst;
        StoreString(inst, len, str);
    }
}
```

There are two important things to note in this example:

1) The `DynamicInterfaceCastableImplementationAttribute` attribute. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL trimming easier, which means NativeAOT scenarios are more reliable.

> 2) The cast to `DemoNativeDynamicWrapper`. This is part of the dynamic nature of `IDynamicInterfaceCastable`. The type that's returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.

Reviewer: > The type that's returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.

It could be my subpar english, but I don't understand what this is trying to say. I know this is about the dynamic castable feature, which is not really com wrappers, so maybe this should link to another doc which explains how this works (I'm curious how this thing works).

Author: Not at all. It is a hard concept and my phrasing isn't great. I tried to explain it in a single sentence and likely didn't do a great job. I will try again.

I think we should have another tutorial on IDIC in general. /cc @elinor-fung

Reviewer 2: Yeah, I'd agree that `IDynamicInterfaceCastable` is complicated and we should have a dedicated tutorial. (All we have is a [sample](https://github.com/dotnet/samples/tree/main/core/interop/IDynamicInterfaceCastable) and [part of a blog post](https://devblogs.microsoft.com/dotnet/improvements-in-native-code-interop-in-net-5-0/#idynamicinterfacecastable) from 5.0)

2) The Native Object Wrapper implements `IDisposable` and calls `Release()` in `Dispose()`.

  > [!NOTE]
  > The `IDisposable` pattern should only be supported if, during the `CreateObject()` call, the `CreateObjectFlags.UniqueInstance` flag was passed in. If this requirement is not followed, it's possible for disposed Native Object Wrappers to be reused after being disposed.

## Using the `ComWrappers` subclass

You now have a `ComWrappers` subclass that can be tested. To avoid creating a native library that returns a COM instance that implements `IDemoGetType` and `IDemoStoreType`, you'll use the Managed Object Wrapper and treat it as a COM instance &ndash; this must be possible in order to pass it COM anyways.

Let's create a Managed Object Wrapper first. Instantiate a `DemoImpl` instance and display its current string state.

```csharp
var demo = new DemoImpl();

string? value = demo.GetString();
Console.WriteLine($"Initial string: {value ?? "<null>"}");
```

Now you can create an instance of `DemoComWrappers` and a Managed Object Wrapper that you can then pass into a COM environment.

```csharp
var cw = new DemoComWrappers();

> IntPtr ccw = cw.GetOrCreateComInterfaceForObject(demo, CreateComInterfaceFlags.None);

Reviewer: I didn't see this sample mention how to handle cases where the code want to `CoCreateInstance` the COM object? I think it might be a relatively common case. Does it need to PInvoke into the Win32 API, or is there some managed helper to make it easier?

Author: >I think it might be a relatively common case.

Agreed. I am going to leave that out. It is conceptually simple once you try it. I don't think it is worth adding that concept here though.

Reviewer: Sorry - does that mean that I should PInvoke into `CoCreateInstance` then? (I agree that it's probably pretty simple to do so, I'm just curious if there's better way)

Author: Yep. Implicit activation has been removed from the runtime but could technically be re-built with `ComWrappers` like C#/WinRT.

```

Instead of passing the Managed Object Wrapper to a COM environment, pretend you just received this COM instance, so you'll create a Native Object Wrapper for it instead.

```csharp

> var rcw = cw.GetOrCreateObjectForComInstance(ccw, CreateObjectFlags.UniqueInstance);

Reviewer: I think both `GetOrCreateComInterfaceForObject` and `GetOrCreateObjectForComInstance` should document their behavior with regard to `AddRef`/`Release` on the native pointer. For example in this case, I honestly can't tell if the code should release `ccw` after it's used here, or if the reference is "Taken over" by the RCW creation. (COM conventions seem to dictate that the `ccw` should be released in this case, so maybe this is a leak?)

**Source generation** &ndash; Most of the code that's needed for COM interop and a `ComWrappers` implementation can likely be autogenerated by some tooling. Source for both types of wrappers could be generated given the proper COM definitions &ndash; for example, Type Library (TLB), IDL, or a Primary Interop Assembly (PIA).

**Global registration** &ndash; Since the `ComWrappers` API was designed as a new phase of COM interop, it needed to have some way to partially integrate with the existing system. There are globally impacting static methods on the `ComWrappers` API that permit registration of a global instance for various support. These methods are designed for `ComWrappers` instances that are expecting to provide comprehensive COM interop support in all cases &ndash; akin to the built-in COM interop system.

> [**Reference Tracker support**][api_referencetracker] &ndash; This support is primary used for WinRT scenarios and represents an advanced scenario. For the majority of `ComWrapper` implementations, either a `CreateComInterfaceFlags.TrackerSupport` or `CreateObjectFlags.TrackerObject` flag should throw a <xref:System.NotSupportedException>. If you'd like to enable this support, perhaps on a Windows or even non-Windows platform, it is highly recommended to reference the [C#/WinRT tool chain][repo_cswinrt].

Reviewer: >  or even non-Windows platform

Does this mean that COM Wrappers work on all platforms? I think that should be mentioned somewhere in the doc. Actually it should mention it either way - whether it's Windows specific or not.

Author: >Actually it should mention it either way - whether it's Windows specific or not.

That is part of the API definition. It is cross platform now. I believe we default to "it is cross-platform unless otherwise stated" &ndash; that seems to be the .NET promise. We updated the API metadata in .NET 6 for cross-platform.

Reviewer: It is surprising to me (in a good way). Mainly because I always think about COM as Windows-only. This is definitely COM, so I expected it to be Windows only. That's why I think it's worth pointing it out.

---
title: "Using the ComWrappers API"
description: A tutorial on properly implementing the ComWrappers API.
author: AaronRobinsonMSFT

> ms.date: "02/09/2021"

Author: ```suggestion
ms.date: "09/02/2021"
```

```csharp
class DemoNativeStaticWrapper
    : IDemoGetType
    , IDemoStoreType
{
    public string? GetString() =>
        throw new NotImplementedException();

    public void StoreString(int len, string? str) =>
        throw new NotImplementedException();
}
```

Constructing an instance of this class and providing it as a wrapper requires defining some policy. If this type is used as a wrapper it would seem that since it implements both interfaces that the underlying COM instance should implement both interfaces too. Given that we adopting this policy we need to ensure this through calls to [`Marshal.QueryInterface()`][api_marshalqueryinterface] on the COM instance to confirm.

```csharp
int hr = Marshal.QueryInterface(ptr, ref IDemoGetType.IID_IDemoGetType, out IntPtr IDemoGetTypeInst);
if (hr != 0)
{
    return null;
}

hr = Marshal.QueryInterface(ptr, ref IDemoStoreType.IID_IDemoStoreType, out IntPtr IDemoStoreTypeInst);
if (hr != 0)
{
    Marshal.Release(ptr);
    return null;
}

return new DemoNativeStaticWrapper()
{
    IDemoGetTypeInst = IDemoGetTypeInst,
    IDemoStoreTypeInst = IDemoStoreTypeInst
};
```

#### Dynamic Native Object Wrapper

> Dynamic wrappers are more flexible in that they provide a way for types to be query at run time instead of statically. In order to provide this support we will utilize [`IDynamicInterfaceCastable`][api_idynamicinterfacecastable]. Observe that `DemoNativeDynamicWrapper` only implements this interface. The functionality that the interface provides is a chance at run time to determine what type is supported.

Reviewer 3: ```suggestion
Dynamic wrappers are more flexible becaues they provide a way for types to be queried at run time instead of statically. In order to provide this support, you'll utilize [`IDynamicInterfaceCastable`][api_idynamicinterfacecastable]. Observe that `DemoNativeDynamicWrapper` only implements this interface. The functionality that the interface provides is a chance to determine what type is supported at run time.
```

```csharp
internal class DemoNativeDynamicWrapper
    : IDynamicInterfaceCastable
{
    public RuntimeTypeHandle GetInterfaceImplementation(RuntimeTypeHandle interfaceType) =>
        throw new NotImplementedException();

    public bool IsInterfaceImplemented(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented) =>
        throw new NotImplementedException();
}
```

> Let's look at one of the types that `DemoNativeDynamicWrapper` will dynamically support. The `IDataStoreType` is defined below using the default interface methods feature.

Reviewer 3: ```suggestion
Let's look at one of the types that `DemoNativeDynamicWrapper` will dynamically support. The following code defines the `IDataStoreType` type using the *default interface methods* feature.
```

```csharp
[DynamicInterfaceCastableImplementation]
unsafe interface IDemoStoreTypeNativeWrapper : IDemoStoreType
{
    public static void StoreString(IntPtr inst, int len, string? str);

    void IDemoStoreType.StoreString(int len, string? str)
    {
        var inst = ((DemoNativeDynamicWrapper)this).IDemoStoreTypeInst;
        StoreString(inst, len, str);
    }
}
```

> There are two important things to note in this example. The first is `DynamicInterfaceCastableImplementationAttribute`. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL Linking easier which means NativeAOT scenarios are more reliable. The second is the cast to `DemoNativeDynamicWrapper`. This is part of the dynamic nature of `IDynamicInterfaceCastable`. The type that is returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.

Reviewer 3: ```suggestion
There are two important things to note in this example:

1) The `DynamicInterfaceCastableImplementationAttribute` attribute. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL Linking easier, which means NativeAOT scenarios are more reliable.
2) The cast to `DemoNativeDynamicWrapper`. This is part of the dynamic nature of `IDynamicInterfaceCastable`. The type that's returned from `IDynamicInterfaceCastable.GetInterfaceImplementation()` is used to blanket the type that implements `IDynamicInterfaceCastable`.
```

> Regardless of which Native Object Wrapper is used, the ability to invoke functions on a COM instance is needed. The implementation of `IDemoStoreTypeNativeWrapper.StoreString()` can serve as an example employing `unmanaged` C# function pointers.

Reviewer 3: ```suggestion
Regardless of which Native Object Wrapper is used, you need the ability to invoke functions on a COM instance. The implementation of `IDemoStoreTypeNativeWrapper.StoreString()` can serve as an example of employing `unmanaged` C# function pointers.
```

```csharp
public static void StoreString(IntPtr inst, int len, string? str)
{
    IntPtr strLocal = Marshal.StringToCoTaskMemUni(str);
    int hr = ((delegate* unmanaged<IntPtr, int, IntPtr, int>)(*(*(void***)inst + 3 /* IDemoStoreType.StoreString slot */)))(inst, len, strLocal);
    if (hr != 0)
    {
        Marshal.FreeCoTaskMem(strLocal);
        Marshal.ThrowExceptionForHR(hr);
    }
}
```

> The dereferencing of the COM instance to access its vtable implementation warrants examining. The COM ABI defines that the first pointer of an object is to the type's vtable and from there the desired slot can be accessed. Let's assume the address of the COM object is `0x10000`. The first pointer sized value should be the address of the vtable &ndash; in this example `0x20000`. Once we are at the vtable we look for the 4th slot (3 in zero-based indexing) to access the `StoreString()` implementation.

Reviewer 3: ```suggestion
Let's examine the dereferencing of the COM instance to access its vtable implementation. The COM ABI defines that the first pointer of an object is to the type's vtable and, from there, the desired slot can be accessed. Let's assume the address of the COM object is `0x10000`. The first pointer-sized value should be the address of the vtable &ndash; in this example `0x20000`. Once you're at the vtable, you look for the fourth slot (index 3 in zero-based indexing) to access the `StoreString()` implementation.
```

```
COM instance
0x10000  0x20000

VTable for IDemoStoreType
0x20000  <Address of QueryInterface>
0x20008  <Address of AddRef>
0x20010  <Address of Release>
0x20018  <Address of StoreString>
```

> Having the function pointer then allows us to dispatch to that member function on that object by passing the object instance as the fist parameter &ndash; that should look familiar based on the function definitions when implementing the Managed Object Wrapper.

Reviewer 3: ```suggestion
Having the function pointer then allows you to dispatch to that member function on that object by passing the object instance as the first parameter. This pattern should look familiar based on the function definitions of the Managed Object Wrapper implementation.
```

> Once the `CreateObject()` method is implemented the `ComWrappers` subclass will be able to produce Native Object Wrappers for COM instances that implement both `IDemoGetType` and `IDemoStoreType`.

Reviewer 3: ```suggestion
Once the `CreateObject()` method is implemented, the `ComWrappers` subclass will be able to produce Native Object Wrappers for COM instances that implement both `IDemoGetType` and `IDemoStoreType`.
```

```csharp
IntPtr iunk = ...; // Get a COM instance from native code.
object rcw = cw.GetOrCreateObjectForComInstance(ccw, CreateObjectFlags.UniqueInstance);
```

### Step 4 &ndash; Native Object Wrapper lifetime details

> The `ComputeVtables()` and `CreateObject()` implementations discussed some wrapper lifetime details but further details need discussion. This can be a short step but can also significantly increase the complexity of the `ComWrappers` design.

Reviewer 3: ```suggestion
The `ComputeVtables()` and `CreateObject()` implementations covered some wrapper lifetime details, but there are further considerations. While this can be a short step, it can also significantly increase the complexity of the `ComWrappers` design.
```

> Unlike the Managed Object Wrapper, which is controlled by calls to its `AddRef()`/`Release()`, the lifetime of a Native Object Wrapper is non-deterministically handled by the GC. The question here is, when does the Native Object Wrapper call `Release()` on the `IntPtr` that represents the COM instance? There are two general buckets here.

Reviewer 3: ```suggestion
Unlike the Managed Object Wrapper, which is controlled by calls to its `AddRef()` and `Release()` methods, the lifetime of a Native Object Wrapper is nondeterministically handled by the GC. The question here is, when does the Native Object Wrapper call `Release()` on the `IntPtr` that represents the COM instance? There are two general buckets:
```

> 1) The Native Object Wrapper's Finalizer should be responsible for calling the COM instance's `Release()`. This is the only time when it is safe to call this method since at this point it has been correctly determined by the GC there are no other references to the Native Object Wrapper in the .NET runtime. There can be complexity here if one is properly supporting COM Apartments &ndash; see the "Additional considerations" section.

Reviewer 3: ```suggestion
1) The Native Object Wrapper's Finalizer is responsible for calling the COM instance's `Release()` method. This is the only time when it's safe to call this method. At this point, it's been correctly determined by the GC that there are no other references to the Native Object Wrapper in the .NET runtime. There can be complexity here if you're properly supporting COM Apartments; for more information, see the [Additional considerations](#additional-considerations) section.
```

> 2) Have the returned Native Object Wrapper implement `IDisposable` and call `Release()` in `Dispose()`. This pattern should only be supported if during the `CreateObject()` call the `CreateObjectFlags.UniqueInstance` flag was passed in. **Note** if this requirement is not followed it is possible for disposed Native Object Wrappers to be reused after being disposed.

Reviewer 3: ```suggestion
2) The Native Object Wrapper implements `IDisposable` and calls `Release()` in `Dispose()`.

  > [!NOTE]
  > This pattern should only be supported if, during the `CreateObject()` call, the `CreateObjectFlags.UniqueInstance` flag was passed in. If this requirement is not followed, it's possible for disposed Native Object Wrappers to be reused after being disposed.
```

## Using the `ComWrappers` subclass

> We now have a `ComWrappers` subclass that can be tested. In order to avoid creating a native library that returns a COM instance that implements `IDemoGetType` and `IDemoStoreType` we are going to use the Managed Object Wrapper and treat it as a COM instance &ndash; this must be possible in order to pass it COM anyways.

Reviewer 3: ```suggestion
You now have a `ComWrappers` subclass that can be tested. To avoid creating a native library that returns a COM instance that implements `IDemoGetType` and `IDemoStoreType`, you'll use the Managed Object Wrapper and treat it as a COM instance &ndash; this must be possible in order to pass it COM anyways.
```

> Let's create a Managed Object Wrapper first. We will instantiate a `DemoImpl` instance display its current string state.

Reviewer 3: ```suggestion
Let's create a Managed Object Wrapper first. Instantiate a `DemoImpl` instance and display its current string state.
```

```csharp
var demo = new DemoImpl();

string? value = demo.GetString();
Console.WriteLine($"Initial string: {value ?? "<null>"}");
```

> Now we can create an instance of `DemoComWrappers` and create a Managed Object Wrapper that we could then pass into a COM environment.

Reviewer 3: ```suggestion
Now you can create an instance of `DemoComWrappers` and a Managed Object Wrapper that you can then pass into a COM environment.
```

```csharp
var cw = new DemoComWrappers();

IntPtr ccw = cw.GetOrCreateComInterfaceForObject(demo, CreateComInterfaceFlags.None);
```

> Instead of passing the Managed Object Wrapper to a COM environment let's pretend we just received this COM instance and instead create a Native Object Wrapper for it.

Reviewer 3: ```suggestion
Instead of passing the Managed Object Wrapper to a COM environment, pretend you just received this COM instance, so you'll create a Native Object Wrapper for it instead.
```

```csharp
var rcw = cw.GetOrCreateObjectForComInstance(ccw, CreateObjectFlags.UniqueInstance);
```

> With the Native Object Wrapper we should be able to cast it to one of the desired interfaces and use it as a normal managed object. We can examine the `DemoImpl` instance and observe the impact of operations on the Native Object Wrapper that is wrapping a Managed Object Wrapper that is in turn wrapping the managed instance is reflected.

Reviewer 3: ```suggestion
With the Native Object Wrapper, you should be able to cast it to one of the desired interfaces and use it as a normal managed object. You can examine the `DemoImpl` instance and observe the impact of operations on the Native Object Wrapper that's wrapping a Managed Object Wrapper that's in turn wrapping the managed instance.
```

I wasn't sure what the last "is reflected" words were trying to convey, so they may need to be added back.

Author: You captured the idea correctly :)

```csharp
var getter = (IDemoGetType)rcw;
var store = (IDemoStoreType)rcw;

string msg = "hello world!";
store.StoreString(msg.Length, msg);
Console.WriteLine($"Setting string through wrapper: {msg}");

value = demo.GetString();
Console.WriteLine($"Get string through managed object: {value}");

msg = msg.ToUpper();
demo.StoreString(msg.Length, msg.ToUpper());
Console.WriteLine($"Setting string through managed object: {msg}");

value = getter.GetString();
Console.WriteLine($"Get string through wrapper: {value}");
```

> Since our `ComWrapper` subclass was designed to support `CreateObjectFlags.UniqueInstance` we can clean up the Native Object Wrapper immediately instead of waiting for a GC to occur.

Reviewer 3: ```suggestion
Since your `ComWrapper` subclass was designed to support `CreateObjectFlags.UniqueInstance`, you can clean up the Native Object Wrapper immediately instead of waiting for a GC to occur.
```

[**Reference Tracker support**][api_referencetracker] &ndash; This support is primary used for WinRT scenarios and represents an advanced scenario. For the majority of `ComWrapper` implementations either a `CreateComInterfaceFlags.TrackerSupport` or `CreateObjectFlags.TrackerObject` flags should throw a `NotSupportedException`. If one would like to enable this support, perhaps on a Windows or even non-Windows platform, it is highly recommended to reference the [C#/WinRT tool chain][repo_cswinrt].

Aside from the lifetime, type system, and functional features that are discussed above, a COM compliant implementation of `ComWrappers` requires additional considerations. The following should also be considered for any implementation that will be used on the Windows platform.

> [**Apartments**][doc_comapartments] &ndash; COM's organizational structure for threading is called "Apartments" and has strict rules that must be followed for stable operations. This tutorial does not implement apartment aware Native Object Wrappers but for any production ready implementation it must. In order to accomplish this we recommend using the [`RoGetAgileReference`][api_rogetagilereference] API introduced in Windows 8. Prior to Windows 8 needs should consider the [Global Interface Table][doc_globalinterfacetable].

Reviewer 3: ```suggestion
- [**Apartments**][doc_comapartments] &ndash; COM's organizational structure for threading is called "Apartments" and has strict rules that must be followed for stable operations. This tutorial does not implement apartment-aware Native Object Wrappers, but any production ready implementation should be apartment-aware. To accomplish this, we recommend using the [`RoGetAgileReference`][api_rogetagilereference] API introduced in Windows 8. For versions prior to Windows 8, consider the [Global Interface Table][doc_globalinterfacetable].
```

> [**Security**][doc_comsecurity] &ndash; COM provides a rich security model for class activation and proxied permission.

Reviewer 3: ```suggestion
- [**Security**][doc_comsecurity] &ndash; COM provides a rich security model for class activation and proxied permission.
```

<!-- Reusable links -->

[api_allocatetypeassociatedmemory]:/dotnet/api/system.runtime.compilerservices.runtimehelpers.allocatetypeassociatedmemory
[api_comwrappers]:/dotnet/api/system.runtime.interopservices.comwrappers
[api_dynamicinterfacecastableimplementation]:/dotnet/api/system.runtime.interopservices.dynamicinterfacecastableimplementationattribute
[api_idynamicinterfacecastable]:/dotnet/api/system.runtime.interopservices.idynamicinterfacecastable
[api_marshalqueryinterface]:/dotnet/api/system.runtime.interopservices.marshal.queryinterface
[api_referencetracker]:/windows/win32/api/windows.ui.xaml.hosting.referencetracker/
[api_rogetagilereference]:/windows/win32/api/combaseapi/nf-combaseapi-rogetagilereference
[api_unmanagedcallersonly]:/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute

[doc_comapartments]:/windows/win32/com/processes--threads--and-apartments
[doc_comsecurity]:/windows/win32/com/security-in-com
[doc_globalinterfacetable]:/windows/win32/com/when-to-use-the-global-interface-table
[doc_impliunknown]:/windows/win32/com/using-and-implementing-iunknown

> [doc_nullable]:/dotnet/csharp/nullable-references

Reviewer 3: ```suggestion
[doc_nullable]:../../csharp/nullable-references.md
```

These relative links are verified by the build.

> [doc_unloadability]:/dotnet/standard/assembly/unloadability

Reviewer 3: ```suggestion
[doc_unloadability]:../assembly/unloadability.md
```

> [doc_unsafekeyword]:/dotnet/csharp/language-reference/keywords/unsafe

Reviewer 3: ```suggestion
[doc_unsafekeyword]:../../csharp/language-reference/keywords/unsafe.md
```

if (hr != 0)
{
    return null;
}

hr = Marshal.QueryInterface(ptr, ref IDemoStoreType.IID_IDemoStoreType, out IntPtr IDemoStoreTypeInst);
if (hr != 0)
{
    Marshal.Release(ptr);
    return null;
}

return new DemoNativeStaticWrapper()
{
    IDemoGetTypeInst = IDemoGetTypeInst,
    IDemoStoreTypeInst = IDemoStoreTypeInst
};
```

#### Dynamic Native Object Wrapper

Dynamic wrappers are more flexible because they provide a way for types to be queried at run time instead of statically. In order to provide this support, you'll utilize [`IDynamicInterfaceCastable`][api_idynamicinterfacecastable]. Observe that `DemoNativeDynamicWrapper` only implements this interface. The functionality that the interface provides is a chance to determine what type is supported at run time.

```csharp
internal class DemoNativeDynamicWrapper
    : IDynamicInterfaceCastable
{
    public RuntimeTypeHandle GetInterfaceImplementation(RuntimeTypeHandle interfaceType) =>
        throw new NotImplementedException();

    public bool IsInterfaceImplemented(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented) =>
        throw new NotImplementedException();
}
```

> Let's look at one of the types that `DemoNativeDynamicWrapper` will dynamically support. The following code defines the `IDataStoreType` type using the *default interface methods* feature.

Reviewer 4: ```suggestion
Let's look at one of the interfaces that `DemoNativeDynamicWrapper` will dynamically support. The following code provides the implementation of `IDemoStoreType` using the *default interface methods* feature.
```

```csharp
[DynamicInterfaceCastableImplementation]
unsafe interface IDemoStoreTypeNativeWrapper : IDemoStoreType
{
    public static void StoreString(IntPtr inst, int len, string? str);

    void IDemoStoreType.StoreString(int len, string? str)
    {
        var inst = ((DemoNativeDynamicWrapper)this).IDemoStoreTypeInst;
        StoreString(inst, len, str);
    }
}
```

There are two important things to note in this example:

> 1) The `DynamicInterfaceCastableImplementationAttribute` attribute. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL Linking easier, which means NativeAOT scenarios are more reliable.

Reviewer 4: ```suggestion
1) The `DynamicInterfaceCastableImplementationAttribute` attribute. This attribute is required on any type that is returned from a `IDynamicInterfaceCastable` method. It has the added benefit of making IL trimming easier, which means NativeAOT scenarios are more reliable.
```

```
COM instance
0x10000  0x20000

VTable for IDemoStoreType
0x20000  <Address of QueryInterface>
0x20008  <Address of AddRef>
0x20010  <Address of Release>
0x20018  <Address of StoreString>
```

Having the function pointer then allows you to dispatch to that member function on that object by passing the object instance as the first parameter. This pattern should look familiar based on the function definitions of the Managed Object Wrapper implementation.

Once the `CreateObject()` method is implemented, the `ComWrappers` subclass will be able to produce Native Object Wrappers for COM instances that implement both `IDemoGetType` and `IDemoStoreType`.

```csharp
IntPtr iunk = ...; // Get a COM instance from native code.

> object rcw = cw.GetOrCreateObjectForComInstance(ccw, CreateObjectFlags.UniqueInstance);

Reviewer 4: ```suggestion
object rcw = cw.GetOrCreateObjectForComInstance(iunk, CreateObjectFlags.UniqueInstance);
```

```

With the Native Object Wrapper, you should be able to cast it to one of the desired interfaces and use it as a normal managed object. You can examine the `DemoImpl` instance and observe the impact of operations on the Native Object Wrapper that's wrapping a Managed Object Wrapper that's in turn wrapping the managed instance.

```csharp
var getter = (IDemoGetType)rcw;
var store = (IDemoStoreType)rcw;

string msg = "hello world!";
store.StoreString(msg.Length, msg);
Console.WriteLine($"Setting string through wrapper: {msg}");

value = demo.GetString();
Console.WriteLine($"Get string through managed object: {value}");

msg = msg.ToUpper();
demo.StoreString(msg.Length, msg.ToUpper());
Console.WriteLine($"Setting string through managed object: {msg}");

value = getter.GetString();
Console.WriteLine($"Get string through wrapper: {value}");
```

Since your `ComWrapper` subclass was designed to support `CreateObjectFlags.UniqueInstance`, you can clean up the Native Object Wrapper immediately instead of waiting for a GC to occur.

```csharp
(rcw as IDisposable)?.Dispose();
```

## Additional considerations

> **Native AOT** &ndash; Ahead of Time (AOT) compilation provides improved startup cost as JIT compilation is avoided. Removing the need for JIT compilation is also often required on some platforms. Supporting AOT was a goal of the `ComWrappers` API, but any wrapper implementation must be careful not to inadvertantly introduce cases where AOT breaks down, such as using reflection. The `Type.GUID` property is an example of where reflection is used, but in a non-obvious way. The `Type.GUID` property uses reflection to inspect the type's attributes and then potentially the type's name and containing assembly in order to generate its value.

Reviewer 2: ```suggestion
**Native AOT** &ndash; Ahead-of-time (AOT) compilation provides improved startup cost as JIT compilation is avoided. Removing the need for JIT compilation is also often required on some platforms. Supporting AOT was a goal of the `ComWrappers` API, but any wrapper implementation must be careful not to inadvertantly introduce cases where AOT breaks down, such as using reflection. The `Type.GUID` property is an example of where reflection is used, but in a non-obvious way. The `Type.GUID` property uses reflection to inspect the type's attributes and then potentially the type's name and containing assembly in order to generate its value.
```
